tests/cloudlaunchprocess/test.csv

```diff
@@ -1,27 +1,25 @@
+run,0,0,q,.sys.kill "test2",,,"Kill the process"
+true,0,0,q,not h in key[.z.W],,,"Check that the handle to the process is no longer open"
+true,0,0,q,not pid1 in "I"$1 _ .proc.sys "ps -u $USER | awk '{print $1}'",,,"Check that the PID of the old process is no longer in use"
 
 comment,0,,"Test 2 - Launch and then kill a process with no extra cmd line args, to test defaulting"
-run,0,0,q,launch input2,,,"Start a new process with no extra command line args, hence relying on defaulting in cloudutils.q"
-run,0,0,q,.proc.sys "sleep 2",,,"Wait for process to start"
-run,0,0,q,"h2:hopen `$""::"",string[discport],"":admin:admin""",,,"Connect to discovery1"
-run,0,0,q,"proc_port:""I""$(neg count string pid)#.Q.s1 h2""exec hpup from .servers.SERVERS where not null w,proctype=`wdb""",,,"Find the process port via discovery"
-run,0,0,q,"h3:hopen `$""::"",string[proc_port],"":admin:admin""",,,"Attempt to connect to the new process"
-true,0,0,q,`default=h3".wdb.writedownmode",,,"Check that connection has been made"
-true,0,0,q,all value input2 in 2 _ (2 cut h3".z.X")[;1],,,"Check that correct process name and type are present in the cmd line args of the process"
-true,0,0,q,"all (getenv[`KDBAPPCONFIG],""/passwords/accesslist.txt"";string .proc.localtime;""0W"") in 2 _ (2 cut h3"".z.X"")[;1]",,,"Check that the expected default parameters for the U,localtime and p flags are present in the cmd line args"
-run,0,0,q,pid:"I"$-2 _ raze raze string a where "q" in' a:.proc.sys "pgrep -lf wdb1",,,"Find the PID of the launched process"
-run,0,0,q,kill "wdb1",,,"Kill the process"
-true,0,0,q,@[{h4 x};".wdb.writedownmode";{1b}],,,"Check that the handle to the process is no longer open"
-true,0,0,q,not first "q" in' b:.proc.sys "pgrep -lf wdb1",,,"Check that the PID of the old process is no longer in use"
+run,0,0,q,.sys.launch input2,,,"Start a new process with no extra command line args, hence relying on defaulting in cloudutils.q"
+run,0,0,q,system "sleep 2",,,"Wait for process to start"
+run,0,0,q,proc_port:first exec hpup from .servers.querydiscovery`test where procname=`test3,,,"Find process port from discovery services "
+run,0,0,q,"h2:hopen `$string[proc_port],"":admin:admin""",,,"Attempt to connect to the process"
+true,0,0,q,`test3=@[h2;".proc.procname";`],,,"Check that connection has been made"
+true,0,0,q,110b~(value input2)in' value key[input2]#.Q.opt h2".z.X",,,"To be reworked:Check that correct process name and type are present in the cmd line args of the process"
+true,0,0,q,"all (getenv[`KDBAPPCONFIG],""/passwords/accesslist.txt"";string .proc.localtime;""0W"")in' value `U`localtime`p#.Q.opt h2"".z.X""",,,"Check that the expected default parameters for the U,localtime and p flags are present in the cmd line args"
+run,0,0,q,pid2:h2".z.i",,,"Find the PID of the launched process"
+run,0,0,q,.sys.kill "test3",,,"Kill the process"
+true,0,0,q,not h2 in key[.z.W],,,"Check that the handle to the process is no longer open"
```

Comment:
I've made all the recommended changes and everything is working except this line (and line 10, which is very similar).
 
The test gives an error when I run the run.sh script but if I test it manually in the same debugger session, just after the tests are finished, it works fine.

The last time something like this happened it was because there was a comma in the code cell, meaning that the csv was being misinterpreted. But this line is so simple I don't know what could be causing it.

Reply:
It gives an error, or the test fails? If it gives an error, I think if you add `-stop` to the command line when starting the test process (i.e. add that inside run.sh), it should drop you into debug mode at the point where it errors and you can poke around.

If it's just the test failing, you might just need to add a sleep after the kill to give time for the other proc to die

Reply:
It gives an error, rather than just failing to return true. And yes I tried adding in a sleep statement after the .sys.kill. Ok I'll use that -stop command thanks

Reply:
I wasn't sure before but looking at the error in debug mode it seems that when the test2/test3 processes are killed the handles h/h2 stay in .z.W, even if they are no longer valid. 

hclose must be used to remove them from.z.W, but this doesn't reflect whether or not the test processes are still runnning. 

As a result I think I'll just keep the previous test: @[h;".proc.procname";{1b}] if that's ok?

Reply:
Hmm, that seems strange. OK sure